packages/js/product-editor/src/blocks/taxonomy/create-taxonomy-modal.tsx

```diff
@@ -0,0 +1,190 @@
+import {
+	useState,
+	useEffect,
+	createElement,
+	createInterpolateElement,
+	useCallback,
+} from '@wordpress/element';
+import { useDispatch } from '@wordpress/data';
+import {
+	__experimentalSelectTreeControl as SelectTree,
+	TreeItemType as Item,
+} from '@woocommerce/components';
+import { useDebounce, useInstanceId } from '@wordpress/compose';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import { Taxonomy } from './types';
+import useTaxonomySearch from './use-taxonomy-search';
+
+type CreateTaxonomyModalProps = {
+	initialName?: string;
+	hierarchical: boolean;
+	slug: string;
+	title: string;
+	onCancel: () => void;
+	onCreate: ( taxonomy: Taxonomy ) => void;
+};
+
+export const CreateTaxonomyModal: React.FC< CreateTaxonomyModalProps > = ( {
+	onCancel,
+	onCreate,
+	initialName,
+	slug,
+	hierarchical,
+	title,
+} ) => {
+	const [ categoryParentTypedValue, setCategoryParentTypedValue ] =
+		useState( '' );
+	const [ allEntries, setAllEntries ] = useState< Taxonomy[] >( [] );
+
+	const { searchEntity, isResolving } = useTaxonomySearch( slug, {
+		fetchParents: false,
+	} );
+
+	const searchDelayed = useDebounce(
+		useCallback(
+			( val ) => searchEntity( val || '' ).then( setAllEntries ),
+			[]
+		),
+		150
+	);
+
+	useEffect( () => {
+		searchDelayed( '' );
+	}, [] );
+
+	const { saveEntityRecord } = useDispatch( 'core' );
+	const [ isCreating, setIsCreating ] = useState( false );
+	const [ errorMessage, setErrorMessage ] = useState< string | null >( null );
+	const [ name, setName ] = useState( initialName || '' );
+	const [ parent, setParent ] = useState< Taxonomy | null >( null );
+
+	const onSave = async () => {
+		setErrorMessage( null );
+		try {
+			const newTaxonomy: Taxonomy = await saveEntityRecord(
+				'taxonomy',
+				slug,
+				{
+					name,
+					parent: parent ? parent.id : null,
+				},
+				{
+					throwOnError: true,
+				}
+			);
+			onCreate( newTaxonomy );
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		} catch ( e: any ) {
+			setIsCreating( false );
+			if ( e.message ) {
+				setErrorMessage( e.message );
+			} else {
+				setErrorMessage(
+					__( `Failed to create taxonomy`, 'woocommerce' )
+				);
+				throw e;
+			}
+		}
+	};
+
+	const id = useInstanceId( BaseControl, 'taxonomy_name' ) as string;
+
+	const selectId = useInstanceId(
+		SelectTree,
+		'parent-taxonomy-select'
+	) as string;
+
+	return (
+		<Modal
+			title={ title }
+			onRequestClose={ onCancel }
+			className="woocommerce-create-new-taxonomy-modal"
+		>
+			<div className="woocommerce-create-new-taxonomy-modal__wrapper">
+				<BaseControl
+					id={ id }
+					label={ __( 'Name', 'woocommerce' ) }
+					help={ errorMessage }
+					className={ classNames( {
+						'has-error': errorMessage,
+					} ) }
+				>
+					<TextControl
+						id={ id }
+						value={ name }
+						onChange={ setName }
+					/>
+				</BaseControl>
+				{ hierarchical && (
+					<SelectTree
```

Comment:
This select tree does not show parents nested, as the categories modal does. It shows them in a flat list, which is confusing....

<img width="680" alt="Screenshot 2023-08-30 at 19 53 31" src="https://github.com/woocommerce/woocommerce/assets/2098816/ed666287-7f69-48ba-9011-7814f5191870">

<img width="680" alt="Screenshot 2023-08-30 at 19 54 20" src="https://github.com/woocommerce/woocommerce/assets/2098816/59d6dd76-8f56-4348-b664-245156e6d434">